gensim/similarities/levenshtein.py

```diff
@@ -8,146 +8,101 @@
+    max_distance : int
+        The maximum Levenshtein distance of the most similar terms.
+        Keeping this value below 3 has a significant impact on the
+        retrieval performance.
 
     See Also
     --------
-    :func:`gensim.similarities.levenshtein.levsim`
-        The Levenshtein similarity.
+    :class:`~gensim.similarities.termsim.WordEmbeddingSimilarityIndex`
+        Retrieve most similar terms for a given term using the cosine similarity over word
+        embeddings.
     :class:`~gensim.similarities.termsim.SparseTermSimilarityMatrix`
         Build a term similarity matrix and compute the Soft Cosine Measure.
 
+    References
+    ----------
+    The Levenshtein similarity in the context of term similarity was defined
+    by [charletetal17]_.
+
+    .. [charletetal17] Delphine Charlet and Geraldine Damnati, "SimBow at SemEval-2017 Task 3:
+       Soft-Cosine Semantic Similarity between Questions for Community Question Answering", 2017,
+       https://www.aclweb.org/anthology/S17-2051/.
+
     """
-    def __init__(self, dictionary, alpha=1.8, beta=5.0, threshold=0.0):
+    def __init__(self, dictionary, alpha=1.8, beta=5.0, max_distance=1):
         self.dictionary = dictionary
         self.alpha = alpha
         self.beta = beta
-        self.threshold = threshold
+        self.max_distance = max_distance
+
+        from lexpy.dawg import DAWG
+
+        self.index = DAWG()
+        terms = sorted(self.dictionary.values())
+        self.index.add_all(terms)
+        self.index.reduce()
+
         super(LevenshteinSimilarityIndex, self).__init__()
 
+    def _levsim(self, t1, t2):
+        from Levenshtein import distance
+
+        max_lengths = max(len(t1), len(t2)) or 1
+        similarity = self.alpha * (1.0 - distance(t1, t2) * 1.0 / max_lengths)**self.beta
+        return similarity
+
     def most_similar(self, t1, topn=10):
-        similarities = (
-            (levsim(t1, t2, self.alpha, self.beta, self.threshold), t2)
-            for t2 in self.dictionary.values()
-            if t1 != t2
-        )
-        most_similar = (
-            (t2, similarity)
-            for (similarity, t2) in sorted(similarities, reverse=True)
-            if similarity > 0
-        )
-        return itertools.islice(most_similar, int(topn))
+        if self.max_distance == 0:
+            terms = []
+        else:
+            effective_topn = topn + 1 if t1 in self.dictionary.token2id else topn
+            for max_distance in range(1, self.max_distance + 1):
+                terms = self.index.search_within_distance(t1, max_distance)
+                if len(terms) >= min(len(self.dictionary), effective_topn):
+                    break
+        most_similar = ((t2, self._levsim(t1, t2)) for t2 in terms if t1 != t2)
+        most_similar = ((t2, similarity) for t2, similarity in most_similar if similarity > 0.0)
+        most_similar = sorted(most_similar, key=lambda x: (x[1], x[0]), reverse=True)
```

Comment:
```suggestion
        most_similar = sorted(most_similar, key=lambda x: (-x[1], x[0]))
```

To avoid sorting terms with the same similarity in reverse lexicographic order (put A before Z, not Z before A).